python/lsst/meas/base/_id_generator.py

```diff
@@ -0,0 +1,548 @@
+    something this:
+
+        from lsst.afw.image import SourceTable, SourceCatalog
+        from lsst.meas.base import VisitIdGeneratorConfig
+        from lsst.obs.base import Instrument
+        from lsst.pex.config import Config from lsst.pipe.base import Task
+
+        class SomeTaskConfig(Config):
+            id_generator = VisitIdGeneratorConfig.make_field()
+
+        class SomeTask(Task):
+
+            ConfigClass = SomeTaskConfig
+
+            ...
+
+            def run(self, ..., data_id: DataCoordinate):
+                id_generator = self.config.apply(data_id)
+                catalog = id_generator.make_source_catalog(self.schema) ...
+
+    Code that wishes to instead unpack these record IDs to obtain the release
+    ID, data ID and counter value should use the same config (often loaded from
+    the ``Butler``) and pass a fully-expanded data ID identifying only a
+    particular ``skymap`` or ``instrument`` to `unpacker_from_config`::
+
+        config = butler.get("some_task_config")
+        catalog = butler.get("some_output_catalog", given_data_id)
+        unpacker = IdGenerator.unpacker_from_config(
+            config.id_generator, butler.registry.expandDataId(skymap="HSC"),
+        )
+        release_id, embedded_data_id, counter = unpacker(catalog[0]["id"])
+        assert embedded_data_id == given_data_id
+
+    This example is a bit contrived, as the ability to reconstruct the data ID
+    is really only useful when you don't have it already, such as when the
+    record ID is obtained from some further-processed version of the original
+    table (such as a SQL database), and in that context the right config to
+    load will not be obvious unless it has been carefully documented.
+
+    Simple instances of the base class that do not include a data ID may also
+    be constructed by calling the constructor directly::
+
+        id_generator = IdGenerator()
+
+    These IDs may not be unpacked, but they also don't need to be, because
+    they're just the per-catalog "counter" integer already.
+    """
+
+    # No deprecation decorator here because the type this method accepts is
+    # itself deprecated, so it's only going to be called be code paths that
+    # will go away when the deprecation turns into a removal, and which already
+    # warn.
+    @staticmethod
+    def _from_exposure_id_info(exposure_id_info: ExposureIdInfo) -> IdGenerator:
+        """Construct a new ID generator from the object this class supersedes.
+
+        This method is deprecated along with the type it accepts; it's provided
+        only as a temporary helper to aid in the transition from
+        `lsst.obs.base.ExposureIdInfo` to `IdGenerator`.
+        """
+        return _ExposureIdInfoIdGenerator(exposure_id_info)
+
+    @property
+    def catalog_id(self) -> int:
+        """The integer identifier for the full catalog with this data ID, not
+        just one of its rows (`int`).
```

Comment:
As I dig into the uses of this, I am surprised that `exposureIdInfo.expId` is replaced with `idGenerator.catalog_id`. Thinking about it, the latter makes sense (you're using it for catalog-y things, not exposure-y things), but it might surprise others when they go to look at it ("why doesn't the new packer have an exposure_id I can get?"). Probably just worth being clear about this change when you announce the merging of this ticket.